### PR TITLE
Implement root tasks

### LINKED ISSUE #845: Run tasks from root directory
### Describe the feature you'd like to request

As a user, I want to be able to cache tasks defined in the root `package.json` and cache them.

```jsonc
{
   "pipeline": {
      "_root#build": {
          "inputs": ["list/of/inputs/to/hash/**"],
          "outputs": ["some/output/dir/**"],
          "dependsOn":["prepare"]
      },
      "_root#prepare": {
          "inputs": ["foo/**"],
          "outputs": ["another/output/dir/**"],
      },
      ....
}      
```

### Describe the solution you'd like

-  https://github.com/vercel/turborepo/issues/523
	- Consider migrating env vars to `inputs` and out of `dependsOn`? idk? 
- Update task graph construction to include root task edges in scheduler https://github.com/vercel/turborepo/blob/main/cli/internal/core/scheduler.go#L183
  - We may also need to adjust resolving root deps, perhaps we can simply add the root to `PackageInfos` and clean up specialized lockfile resolution https://github.com/vercel/turborepo/blob/main/cli/internal/context/context.go#L164

### Describe alternatives you've considered

None

### PR DESCRIPTION
**Pull request by gsoltis**
Adds the ability to run tasks from the root `package.json` file.

Implementation notes that will need to be documented:
 * Tasks run from the root are opt-in. That is, `turbo.json`*must* have an entry for `_root#<task>` to enable running that task at the root.
 * `_root` is available in filters, so `turbo run my-task --filter=_root` will run `_root#my-task`. 
 * If an entry exists in `turbo.json` for `_root#task`, then `turbo run task` will also run that task for the root.
 * There is some best-effort cycle detection in place. If the root `package.json` has `"build": "turbo run build"`, then even after opting in, `turbo run build --filter=_root` will produce an error. This cycle detection can be circumvented with aliases or other indirection used to run `turbo`.
 * For the purposes of changed files, any changed file in a repo that doesn't belong to another package is considered to belong to the root.
   * Note that this could cause frequent runs of root tasks if they are triggered from git changes. With `inputs` specified, they should hit cache if nothing relevant has changed.

Additional changes:
 * `e2e` tests now include a root task
 * a package task is now sufficient to mark a particular task as runnable by `turbo`. e.g. if you defined `foo#my-task` but not `my-task` in `turbo.json`, previously running `turbo run my-task` would be an error, now it will run `foo#my-task`.

Fixes #845

### CODE REVIEW

File: cli/internal/scope/filter/filter.go [hunk 1]
```diff
@@ -202,7 +202,16 @@ func (r *Resolver) filterNodesWithSelector(selector *TargetSelector) (util.Set,
 		parentDir := selector.parentDir
 		for pkgName := range changedPkgs {
 			if parentDir != "" {
-				if pkg, ok := r.PackageInfos[pkgName]; !ok {
+				if pkgName == util.RootPkgName {
+					// The root package changed, only add it if
+					// the parentDir is equivalent to the root
+					//if parentDir == r.Cwd {
+					if matches, err := doublestar.PathMatch(parentDir, r.Cwd); err != nil {
+						return nil, fmt.Errorf("failed to resolve directory relativeeship %v contains %v: %v", parentDir, r.Cwd, err)
```
Comment (jaredpalmer) [inline cli/internal/scope/filter/filter.go:210]: Fix error message here

File: cli/internal/scope/filter/filter.go [hunk 2]
```diff
@@ -202,7 +202,16 @@ func (r *Resolver) filterNodesWithSelector(selector *TargetSelector) (util.Set,
 		parentDir := selector.parentDir
 		for pkgName := range changedPkgs {
 			if parentDir != "" {
-				if pkg, ok := r.PackageInfos[pkgName]; !ok {
+				if pkgName == util.RootPkgName {
+					// The root package changed, only add it if
+					// the parentDir is equivalent to the root
+					//if parentDir == r.Cwd {
```
Comment (jaredpalmer) [inline cli/internal/scope/filter/filter.go:208]:
```suggestion
```

**Review by jaredpalmer — CHANGES_REQUESTED** [review]

**Comment by gsoltis** [comment]
Windows `e2e` test is still broken, so not quite ready for merge yet

**Comment by foogazy** [comment]
Hello! This would be huge! How can we help for this to be merged?

**Comment by gsoltis** [comment]
@foogazy thanks for your interest!

This is coming soon, although there may be some slight tweaks. 

The most helpful thing would be to verify that it works for your use cases.

**Comment by foogazy** [comment]
@gsoltis Hello! Thank you so much for all your amazing work! 

Sure! I tested on my system at can verify it works fine: 

System details: 

- Macbook Pro - MacOS Monterey - M1 Chip - 16GB Memory 
- Node v. 16 
- Yarn 1.22.18
- npm 8.3.1

Hope this helps. :)

**Comment by gsoltis** [comment]
@foogazy Awesome, glad to hear it!

**Comment by dagda1** [comment]
Oh man, I wish this was merged.

**Comment by trevorr** [comment]
@gsoltis Does this support referencing root package-tasks in `dependsOn`? When I try that with v1.3.1, I get `error preparing engine: found reference to unknown package: //`.

**Pull request merged**